src/jit/simdcodegenxarch.cpp

```diff
@@ -2148,8 +2148,10 @@ void CodeGen::genSIMDIntrinsicRelOp(GenTreeSIMD* simdNode)
                 getEmitter()->emitIns_R_I(INS_cmp, EA_4BYTE, intReg, mask);
             }
 
-            if (targetReg != REG_NA)
+            if ((simdNode->gtFlags & GTF_SET_FLAGS) == 0)
```

Comment:
I would add an else clause here, asserting that `targetReg == REG_NA`. I would actually be more inclined to reverse the sense of this. That is, I would check whether `targetReg != REG_NA`, and then assert that `GTF_SET_FLAGS` is not set/set in the if and else clause, but I guess they are basically equivalent.

Reply:
Yes, the existing `if` should stay as is but I have problems getting this to work because the RA insists on allocating a register even if `dstCount` is 0. Need to look into this a bit more.

Reply:
This is caused by this piece of code in lsra.cpp
https://github.com/dotnet/coreclr/blob/10c320c512e5246b658c76c1c699233e1ec8a35d/src/jit/lsra.cpp#L4821-L4828

It forces `isLocalDefUse` to `true` for unused value nodes. That's probably intended to cover the common case of x86 instructions (e.g. `add eax, ebx`) but it's not suitable in this situation.

Reply:
Ah, I *think* this is similar to compares, where I had to add an `isNoRegCompare` bit to the `TreeNodeInfo` to handle the case where you've got a compare that you don't want to allocate a register for.

Reply:
I looked into `isNoRegCompare` before but it's purpose seems to be rather different, it's used by "Contain" code to communicate to "TreeNodeInfoInit" code that `dstCount` should be 0. But LSRA itself doesn't appear to use `isNoRegCompare` in any way so setting it on the SIMD node has no effect.

What's not clear to me is why LSRA forces `isLocalDefUse` to true based on `IsValue`/`IsUnusedValue` instead of relying on the information provided by `TreeNodeInfoInit`.

Reply:
Also note that the `TreeNodeInfoInit` function LSRA calls ends with the following piece of code:
https://github.com/dotnet/coreclr/blob/43cf34fe74b435427ffb62f54fc730b4b46c4c22/src/jit/lsraxarch.cpp#L719-L725

This is very similar to the code I quoted above yet slightly different. Are these 2 pieces of code correct?

Reply:
> What's not clear to me is why LSRA forces isLocalDefUse to true based on IsValue/IsUnusedValue instead of relying on the information provided by TreeNodeInfoInit.

Ultimately, the setting of `dstCount` is/should be based on:
- If the node is contained it is 0 (though that's largely irrelevant because the value won't be used)
- If `!IsValue()` it is 0.
- If `IsNoRegCompare()` it is 0
- If `IsValue()` it is 1, or more if it is a node that defines multiple registers

There are places outside of LSRA that need to know the number of registers defined by a node. So the plan is that one should be able to determine the dstCount with `gtLsraInfo`, which is being eliminated. And in my next round of changes, I'm adding an assert at the end of `LinearScan::TreeNodeInfoInit()`:
```
    assert(info->dstCount == tree->GetRegisterDstCount());
```
Where `GetRegisterDstCount()` is a new method that does the above checks.